dftimewolf/lib/args_validator.py

```diff
@@ -355,6 +356,113 @@ def _ValidateOrder(self,
           f'{first} is after {second} but it should be the other way around')
 
 
+class FQDNValidator(RegexValidator):
+  """Validator for FQDNs."""
+
+  NAME = 'fqdn'
+  FQDN_REGEX = (
+      r'(?=^.{4,253}$)(^((?!-)[a-zA-Z0-9-]{1,63}(?<!-)\.)+[a-zA-Z]{2,63}$)')
+  # Source: https://stackoverflow.com/questions/11809631/fully-qualified-domain-name-validation#20204811  # pylint: disable=line-too-long
+  # Retrieved 2023-02-03
+
+  def ValidateSingle(self,
+                     operand: str,
+                     validator_params: Optional[Dict[str, Any]] = None) -> None:
+    """Validate an FQDN.
+
+    Args:
+      operand: The FQDN to validate.
+      validator_params: Unused for this validator.
+
+    Raises:
+      errors.RecipeArgsValidatorError: Raised if argument fails validation.
+    """
+    if not validator_params:
+      validator_params = {}
+    validator_params['regex'] = self.FQDN_REGEX
+
+    try:
+      super().ValidateSingle(operand, validator_params)
+    except errors.RecipeArgsValidatorError:
+      # Give a nicer error message than the regex failure
+      raise errors.RecipeArgsValidatorError(
+          f"'{operand}' is an invalid hostname.")
+
+
+class GRRHostValidator(FQDNValidator):
+  """Validates a GRR hostname.
+
+  Grr can accept FQDNs, or Grr client IDs, which take the form of
+  C.1facf5562db006ad.
+  """
+
+  NAME = 'grr_host'
+  GRR_REGEX = r'^C\.[0-9a-f]{16}$'
+
+  def ValidateSingle(self,
+                     operand: str,
+                     validator_params: Optional[Dict[str, Any]] = None) -> None:
+    """Validate a Grr host ID.
+
+    Args:
+      operand: The ID to validate.
+      validator_params: Unused for this validator.
+
+    Raises:
+      errors.RecipeArgsValidatorError: Raised if argument fails validation.
+    """
+    if not validator_params:
+      validator_params = {}
+
+    # Grr Host IDs can be FQDNs, test that first
+    try:
+      FQDNValidator.ValidateSingle(self, operand, validator_params)
+    except errors.RecipeArgsValidatorError:
+      # Hostname validation failed, check for Grr client ID.
+      validator_params['regex'] = self.GRR_REGEX
```

Comment:
Since this validator is intended to validate a GRR hostname, would it make more sense to check that first and then if that fails, clear validator_params and then call super() again to check the hostname?

That might remove the need to use the ClassName.ValidateSingle(...) calls you are doing here.

Reply:
The order (Grr client ID vs FQDN first) probably doesn't make that much difference, it's arbitrary. 

Also, `ValidateSingle` is indeed the correct method to go with here, as walking up the class inheritance ladder is `CommaSeparatedValidator` which requires `ValidateSingle` rather than `Validate` be overridden.